File: Python/Complex.py
from math import atan, cos, e, pi, sin
class Complex:
	def __init__(self, a=0, b=0, pol=False, disp=False, sym='i'):
		if pol not in (False, True):
			b = b, self.angle(pol)
		elif pol:
			pol = b[1][0]
		self.a, self.b, self.u, self.disp, self.sym = a, b, pol, disp, sym
	def angle(self, u):
		u = u.lower().strip()[:1]
		if u == 'r':
			return ' rad', 1
		if u == 'g':
			return u, 200/pi
		return '°', 180/pi
	def mod(self):
		fm = (self.a**2+self.b**2)**0.5
		m = int(fm)
		if fm == m:
			fm = m
		return fm
	def arg(self, u):
		u = self.angle(u)
		if self.a>0:
			return Complex(b=(atan(self.b/self.a)*u[1], u))
		if self.a==0:
			if self.b>0:
				return Complex(b=(pi/2*u[1], u))
			elif self.b<0:
				return Complex(b=(-pi/2*u[1], u))
			return Complex(b=(0, u))
		if self.b>=0:
			return Complex(b=((atan(self.b/self.a)+pi)*u[1], u))
		return Complex(b=((atan(self.b/self.a)-pi)*u[1], u))
	def conjg(self):
		if self.u != False:
			return Complex(self.a, -self.b[0], self.b[1][0])
		return Complex(self.a, -self.b)
	def reci(self):
		return Complex(1, 0, self.u)/self
	def pol(self, u=''):
		return Complex(self.mod(), self.arg(u).b, pol=True)
	def rec(self):
		i = self.b[0]/self.b[1][1]
		return Complex(self.a*cos(i), self.a*sin(i))
	def __add__(self, z):
		if self.disp:
			if not l:
				l.append(0)
			l[0] += 1
			i, j = l[0], l[0] + 1
		u = False
		if self.u != False or z.u != False:
			u = self.u
			if self.u != False and z.u != False and self.b == (self.b[1][1]/z.b[1][1]*z.b[0], self.b[1]):
				return Complex(self.a+z.a, self.b, pol=True)
			if self.u != False:
				self = self.rec()
				if self.disp:
					print(' + '.join(f"z{k+1}" for k in range(i)), f"= {self}")
			if z.u != False:
				z = z.rec()
				if self.disp:
					print(f"z{j} = {z}")
		z = Complex(self.a+z.a, self.b+z.b)
		if u != False:
			if self.disp:
				print(' + '.join(f"z{k+1}" for k in range(j)), f"= {z}")
			if j == n:
				z = z.pol(u)
		return z
	def __sub__(self, z):
		if self.disp:
			if not l:
				l.append(0)
			l[0] += 1
			i, j = l[0], l[0] + 1
		u = False
		if self.u != False or z.u != False:
			u = self.u
			if self.u != False and z.u !=False and self.b == (self.b[1][1]/z.b[1][1]*z.b[0], self.b[1]):
				return Complex(self.a-z.a, self.b, pol=True)
			if self.u != False:
				self = self.rec()
				if self.disp:
					print(' - '.join(f"z{k+1}" for k in range(i)), f"= {self}")
			if z.u != False:
				z = z.rec()
				if self.disp:
					print(f"z{j} = {z}")
		z = Complex(self.a-z.a, self.b-z.b)
		if u != False:
			if self.disp:
				print(' - '.join(f"z{k+1}" for k in range(j)), f"= {z}")
			if j == n:
				z = z.pol(u)
		return z
	def __mul__(self, z):
		if self.disp:
			if not l:
				l.append(0)
			l[0] += 1
			j = l[0] + 1
		if self.u != False:
			if z.u == False:
				z = z.pol(self.u)
				if self.disp:
					print(f"z{j} = {z}")
			elif self.u != z.u:
				z.b = self.b[1][1]/z.b[1][1]*z.b[0], self.b[1]
				if self.disp:
					print(f"z{j} = {z}")
			return Complex(self.a*z.a, (self.b[0]+z.b[0], self.b[1]), pol=True)
		if z.u != False:
			z = z.rec()
			if self.disp:
				print(f"z{j} = {z}")
		return Complex(self.a*z.a-self.b*z.b, self.a*z.b+self.b*z.a)
	def __truediv__(self, z):
		if self.disp:
			if not l:
				l.append(0)
			l[0] += 1
			j = l[0] + 1
		if self.u != False:
			if z.u == False:
				z = z.pol(self.u)
				if self.disp:
					print(f"z{j} = {z}")
			elif self.u != z.u:
				z.b = self.b[1][1]/z.b[1][1]*z.b[0], self.b[1]
				if self.disp:
					print(f"z{j} = {z}")
			return Complex(self.a/z.a, (self.b[0]-z.b[0], self.b[1]), pol=True)
		if z.u != False:
			z = z.rec()
			if self.disp:
				print(f"z{j} = {z}")
		c = z.a**2+z.b**2
		return Complex((self.a*z.a+self.b*z.b)/c, (z.a*self.b-self.a*z.b)/c)
	def __pow__(self, z):
		if self.u == False:
			self = self.pol()
		if z.u != False:
			z = z.rec()
		return Complex(self.a/e**(self.b[0]/self.b[1][1]*z.b), (z.a*self.b[0], self.b[1]), pol=True)
	def __repr__(self):
		if type(self.b) == tuple:
			s, b, u = '', *self.b
			if self.u != False:
				a = self.a
				x, y = int(a), int(b)
				if x == a:
					a = x
				if y == b:
					b = y
				s = f"{a}∠"
			s += f"{b}{u[0]}"
			return s
		a, b, sym = self.a, self.b, self.sym
		x, y = int(a), int(b)
		if x == a:
			a = x
		if y == b:
			b = y
		if b >= 0:
			if b - 1:
				if b:
					return f"{a}+{b}{sym}"
				return f"{a}"
			return f"{a}+{sym}"
		if b + 1:
			return f"{a}-{-b}{sym}"
		return f"{a}-{sym}"
	def __str__(self):
		return repr(self)
if __name__ == '__main__':
	while 1:
		print('*'*16,' '.join('COMPLEX'),'',' '.join('NUMBER'),'*'*16)
		choices = 'Modulus Argument Conjugate Reciprocal Polar Rectangular Add Subtract Multiply Divide Power Exit'.split()
		for i, j in enumerate(choices):
			print(f"{i+1}. {j}")
		ch = input('Enter your choice :\t')
		if ch in '1 2 3 4 5 6'.split():
			if ch in ['3', '4']:
				pr = input('1. Polar\n2. Rectangular\nEnter your choice :\t')
			else:
				pr = '1' if ch == '6' else '2'
			if pr == '1':
				r = eval(input("Enter magnitude of z :\t").replace('^', '**') or 0)
			elif pr == '2':
				a = eval(input("Enter real part of z :\t").replace('^', '**') or 0)
				b = eval(input("Enter imaginary part of z :\t").replace('^', '**') or 0)
				z = Complex(a, b)
			else:
				print('Invalid choice.')
				continue
			if pr == '1' or ch in ['2', '5']:
				u = input("Enter its angle unit (d/r/g) :\t")
			if pr == '1':
				i = eval(input("Enter its inclination :\t").replace('^', '**') or 0)
				z = Complex(r, i, u)
			print(f"z = {z}")
			if ch == '1':
				z = '|z|', z.mod()
			elif ch == '2':
				z = 'arg(z)', z.arg(u)
			elif ch == '3':
				z = 'Conjg(z)', z.conjg()
			elif ch == '4':
				z = 'z⁻¹', z.reci()
			elif ch == '5':
				z = 'Pol(z)', z.pol(u)
			else:
				z = 'Rec(z)', z.rec()
			print(f"{z[0]} = {z[1]}")
		elif ch in '7 8 9 10 11'.split():
			z = []
			n = int(input(f"{choices[int(ch)-1]} how many? :\t"))
			yn = input('Display in between conversions? (y/n) :\t').lower().strip()[:1] == 'y'
			for j in range(1, n + 1):
				pr = input('1. Polar\n2. Rectangular\nEnter your choice :\t')
				if pr == '1':
					r = eval(input(f"Enter magnitude of z{j} :\t").replace('^', '**') or 0)
					u = input("Enter its angle unit (d/r/g) :\t")
					i = eval(input("Enter its inclination :\t").replace('^', '**') or 0)
					z.append(Complex(r, i, u, disp=yn))
				else:
					a = eval(input(f"Enter real part of z{j} :\t").replace('^', '**') or 0)
					b = eval(input(f"Enter imaginary part of z{j} :\t").replace('^', '**') or 0)
					z.append(Complex(a, b, disp=yn))
			for i, j in enumerate([i for i in z]):
				print(f"z{i+1} = {j}")
			op = dict(zip('7 8 9 10 11'.split(), '+ - * / **'.split()))[ch]
			l = []
			z = f" {op} ".join(f"z{i+1}" for i in range(n)), eval(op.join(f"z[{i}]" for i in range(n)))
			print(f"{z[0]} = {z[1]}")
		elif ch == '12':
			break
		else:
			print('Invalid choice.')

File: complex.cpp
#include <iostream>
#include <math.h>
#include "complex.h"

using namespace std;

char complex::drg;
char *complex::u;
bool complex::p = false;

void complex::angle(bool from_rad)
{
	switch (drg)
	{
	case 'r':
	case 'R':
		break;
	case 'g':
	case 'G':
		y *= from_rad ? 200 / M_PI : M_PI / 200;
		break;
	default:
		y *= from_rad ? 180 / M_PI : M_PI / 180;
	}
}

double complex::convert(double i, char u, bool from_rad)
{
	complex z;
	z.y = i, z.drg = u;
	z.angle(from_rad);
	return z.y;
}

char *complex::unit(char u)
{
	switch (u)
	{
	case 'r':
	case 'R':
		return (char *)" rad";
	case 'g':
	case 'G':
		return (char *)"g";
	default:
		return (char *)"\u00b0";
	}
}

double complex::mod()
{
	return sqrt(x * x + y * y);
}

double complex::arg()
{
	if (x > 0)
		return atan(y / x);
	if (x == 0)
	{
		if (y > 0)
			return M_PI / 2;
		if (y < 0)
			return -M_PI / 2;
		return 0;
	}
	if (y >= 0)
		return atan(y / x) + M_PI;
	return atan(y / x) - M_PI;
}

complex complex::conjugate()
{
	complex z1;
	z1.x = x;
	z1.y = -y;
	z1.p = p;
	z1.u = u;
	return z1;
}

complex complex::reci()
{
	complex z1, z2;
	z1.x = 1, z1.y = 0, z2.x = this->x, z2.y = this->y;
	z1.p = z2.p = this->p, z1.u = z2.u = this->u;
	z1.drg = z2.drg = this->drg;
	return z1 / z2;
}

complex complex::pol(char u)
{
	complex z, p;
	z.x = this->x, z.y = this->y;
	p.x = z.mod(), p.y = complex::convert(z.arg(), u), p.p = true, p.u = complex::unit(u);
	return p;
}

complex complex::rec()
{
	complex z1, z2;
	z1.x = this->x, z1.y = this->y, z1.drg = this->drg;
	z1.angle(false);
	z2.x = z1.x * cos(z1.y), z2.y = z1.x * sin(z1.y);
	return z2;
}

complex complex::operator+(complex &m)
{
	complex z1, z2;
	z1.x = this->x, z1.y = this->y, z1.p = p, z1.u = u;
	if (p || m.p)
	{
		if (p && m.p && y == complex::convert(1, drg) * complex::convert(m.y, m.drg, false))
		{
			z1.x += m.x;
			return z1;
		}
		if (p)
			z1 = z1.rec();
		if (m.p)
			m = m.rec();
	}
	z2.x = z1.x + m.x;
	z2.y = z1.y + m.y;
	if (p)
		z2 = z2.pol(drg);
	return z2;
}

complex complex::operator-(complex &m)
{
	complex z1, z2;
	z1.x = this->x, z1.y = this->y, z1.p = p, z1.u = u;
	if (p || m.p)
	{
		if (p && m.p && y == complex::convert(1, drg) * complex::convert(m.y, m.drg, false))
		{
			z1.x -= m.x;
			return z1;
		}
		if (p)
			z1 = z1.rec();
		if (m.p)
			m = m.rec();
	}
	z2.x = z1.x - m.x;
	z2.y = z1.y - m.y;
	if (p)
		z2 = z2.pol(drg);
	return z2;
}

complex complex::operator*(complex &m)
{
	complex z1;
	if (p)
	{
		if (!m.p)
			m = m.pol(drg);
		else if (drg != m.drg)
			m.y = complex::convert(1, drg) * complex::convert(m.y, m.drg, false), m.u = u;
		z1.x = x * m.x;
		z1.y = y + m.y;
		z1.p = p, z1.u = u;
		return z1;
	}
	if (m.p)
		m = m.rec();
	z1.x = x * m.x - y * m.y;
	z1.y = x * m.y + m.x * y;
	return z1;
}

complex complex::operator/(complex &m)
{
	complex z1;
	double c;
	if (p)
	{
		if (!m.p)
			m = m.pol(drg);
		else if (drg != m.drg)
			m.y = complex::convert(1, drg) * complex::convert(m.y, m.drg, false), m.u = u;
		z1.x = x / m.x;
		z1.y = y - m.y;
		z1.p = p, z1.u = u;
		return z1;
	}
	if (m.p)
		m = m.rec();
	c = m.x * m.x + m.y * m.y;
	z1.x = (x * m.x + y * m.y) / c;
	z1.y = (m.x * y - x * m.y) / c;
	return z1;
}

complex complex::operator^(complex &m)
{
	complex z1, z2;
	z1.x = x, z1.y = y;
	if (!p)
		z1 = z1.pol();
	if (m.p)
		m = m.rec();
	z2.x = z1.x / exp(complex::convert(z1.y, 'd', false) * m.y), z2.y = m.x * z1.y, z2.p = true, z2.u = z1.u;
	return z2;
}

ostream &operator<<(ostream &out, const complex &c)
{
	out << c.x;
	if (c.y > 0)
		out << (c.p ? "\u2220" : "+") << c.y << (c.p ? c.u : "i");
	else if (c.y < 0)
		out << (c.p ? "\u2220" : "") << c.y << (c.p ? c.u : "i");
	return out;
}

istream &operator>>(istream &in, complex &c)
{
	char u;
	cout << (c.p ? "Enter the Magnitude :\t" : "Enter the Real Part :\t");
	in >> c.x;
	if (c.p)
	{
		cout << "Enter its Angle Unit (d/r/g) :\t";
		in >> u;
	}
	cout << (c.p ? "Enter its Inclination :\t" : "Enter the Imaginary Part :\t");
	in >> c.y;
	if (c.p)
		c.drg = u, c.u = complex::unit(u);
	return in;
}